cma/evolution_strategy.py

```diff
@@ -4646,4 +4652,28 @@ def update(es):
     es.objective_function_complements = [_al]
     es.augmented_lagrangian = _al
     es.best_feasible = best_feasible_solution
+
+    if post_optimization:
+        positive_constraints = np.where(np.array(g(es.result.xfavorite)) > 0)
+        if len(positive_constraints[0]) > 0:
+            x_post_opt, es_post_opt = fmin_con(lambda x: np.sum(np.square(np.array(g(x))[positive_constraints])),
```

Comment:
It seems that here only some of the constraints g squared are minimized. I don't see how i) not taking all constraints and ii) optimizing g**2 is likely to end up in an overall feasible solution.

Reply:
From our understanding, the idea was to only minimize g**2 for the constraints that are positive in x_favorite to move closer to the order side of the constraint border while staying close to the solution x_favorite. We only take the positive constraints since they indicate the direction in which the solution needs to move to get feasible. In the fmin_con function called in post-optimization, we keep the same equality and inequality constraints as in the previous optimization to ensure that the solution returned by the post-optimization will be feasible.

Reply:
The problems I see are the following:

- The constraints that are feasible at the start of the post optimization do not necessarily stay feasible, just as the constraints that are now in `f` did not end up feasible in the first place.
- Optimizing `g**2` does neither guaranty feasibility nor even optimize for feasibility. It optimizes for getting close to the boundary of the feasible space. The latter is also what `fmin_con` does in the first place.

The suggestion in the issue "optimizing the sum of the positive constraints squared" was meant to say to optimize for ```sum([gi**2 if gi > 0 else 0 for gi in g(x)])```.

Using `es.sigma` as initial step-size seems a good alternative, or using the `es.optimize` method.

Reply:
I fully understand your two points, and to make sure that the solution returned after post optimization is feasible, we used the best feasible solution returned by the `fmin_con()` function (i.e. `es_post_opt.best_feasible`). However, with `sum([gi**2 if gi > 0 else 0 for gi in g(x)])` as an objective function, we find that the results have very poor consistency and the post-optimization step rarely returns a feasible solution (especially in higher dimensions). To illustrate, I plotted the same graph as in the bottom of this page for sum([gi**2 if gi > 0 else 0 for gi in g(x)])` as an objective function (missing points are due to no feasible solution returned).

For the choice of `sigma0_post_opt = es.sigma`, I do agree that this choice makes the most sense.

![Figure_7](https://user-images.githubusercontent.com/63016186/144311606-0150254c-dbe9-4425-bd7c-e9a6c6b6ef0d.png)

![Figure_8](https://user-images.githubusercontent.com/63016186/144311614-9bf02970-a1fa-459e-9a51-1b7a445b4704.png)